### ISSUE TITLE
Lazy and friends swallow fatal errors during implicit resolution

### ISSUE THREAD

**Issue by joroKr21** [milessabin/shapeless#776]
`c.inferImplicitValue` seems to have quite an extreme opinion on what `silent` means:
```scala
trait Tc[A]
object Tc {
  implicit def materialize[A]: Tc[A] = macro materializeImpl[A]
  def materializeImpl[A](c: blackbox.Context): c.Tree =
    throw new OutOfMemoryError("Whoops, I ran out of memory during implicit resolution.")
}

object UseTc {
  implicitly[Tc[Int]] // out of memory
  implicitly[Lazy[Tc[Int]]] // missing implicit !?
}
```
Now what is the chance this ever happened? Very low.
Did it happen here: https://travis-ci.org/milessabin/shapeless/jobs/274500922? Who knows?

**Comment by joroKr21** [comment]
Or the problem is here: https://github.com/milessabin/shapeless/blob/master/core/src/main/scala/shapeless/lazy.scala#L320-L322

**Comment by milessabin** [comment]
Interesting! Yes, that, looks like it might be an explanation. If it is we should be able to make the 2.10.x problem go away by bumping up the heap and/or stacksize ... other things being equal I'd prefer to leave the `Lazy`'s as they are in the example, at least for the time being.

**Comment by joroKr21** [comment]
#777 passed this time. Hopefully it continues to do so in the future. I don't give sbt more than 2g memory on my machine either.

Apart from that the issue with the swallowed error seems unfixable. `silent = false` just converts every error to "implicit search has failed. to find out the reason, turn on -Xlog-implicits". Very helpful... Maybe this can be fixed in scalac. But I can't find where the error is caught.

**Comment by joroKr21** [comment]
This is hit and miss. If you don't mind me asking, why do you want to keep the `Lazy` [in tail position](https://github.com/milessabin/shapeless/blob/master/examples/src/main/scala/shapeless/examples/sexp.scala#L300)? It does nothing there.

**Comment by macalinao** [comment]
Thanks for reporting this. The solution for me was to add the following to my `SBT_OPTS` env var:

```
-Xmx6G -Xms6G -Xss2M
```

**Comment by joroKr21** [comment]
Wow, 6G is crazy :scream: I opened milessabin/kittens#61

**Comment by macalinao** [comment]
I think it was probably the stack size that fixed the issue. I agree 6G is
overkill.

**Comment by joroKr21** [comment]
Fixed in scalac by scala/scala#6344, nothing to be done here
